### ISSUE TITLE
Addition of test information programmatically

### ISSUE THREAD

**Issue by schhil** [renatoathaydes/spock-reports#123]
I would love to be able to add dynamically computed information to the spock-report for each test case execution.
Example: I have a some hundert REST service tests that do GET-Requests which look somehow like this:
```
when:
      requestHeaders.add("clientname", "webshop")
      get("api/item/{id}", id: idParameter)
then:
      result.statusCode.'2xxSuccessful'
and: ...
where:
      idParameter << [...]
```
My get()-method in this case is implemented in the test base class, which is derived from spock.lang.Specification. It will _dynamically_ construct the complete URL and do the REST service call. So my method could do a call or during test setup I could provide a method, which spock-report would call at the end of the test execution, so that the context information comes to the report.
If the test fails (or even if it does not fail), it would be great to see the context data in the spock-report.
The data structure could be a java.util.map and the representation in the report could be a line by line name/value pair.

**Comment by renatoathaydes** [comment]
I like this idea. This is already possible by writing your own template, but I understand that could be hard to do if you are not familiar with templates.

I might have a look at this in the next few weeks (I am on vacation now, not much time to look).

**Comment by renatoathaydes** [comment]
@schhil I've implemented this feature, could you give it a try by checking out the `next` branch and running your tests locally?

To add information to reports, just do something like this:

```groovy
when: 'something something'
reportInfo "The data you want to be included in the report!"
```

**Comment by perwezalamnewscorp** [comment]
Good to hear that its already implemented. I think this implementation is more of individual test information and its great to have this. And my requirement is about test environment. 
What if i could provide the test environment information in setupSpec, so that  these information can be added into the report. There are many more for information that can be added into the report like, My Test application URL, Test Application name, Test environment as (Dev,Test, Stage, Prod etc) or other information as  well

**For example** 

    def setupSpec(){
       testInfo = {browserName:"chrome", osName:"Mac", testEnvironment:"Stage"}
       testEnvironmentInfo.add(testInfo)
       reportTitle "My Dashboard"

    }

And also I was talking tag information. The tag could be something related to the feature category or name, and priority level of the test

**For example**

Feature name could be a tag as (Login, Home Page, Logout, payment, my Wallet etc) and these info should be able to add individually for all test like you have done as **reportInfo**

**addTag "Login"
addPriority "High"**

**Comment by renatoathaydes** [comment]
Adding info in `setupSpec` seems very useful, should be easy to do based on my changes... but I don't see the point in having `addTag`, `addPriority` etc... you could create your own little library that generates HTML then calls `reportInfo` with that, then it will look whatever way you like in the report.

For the `setupSpec` extra info, I can add something like `reportHeader 'This goes on the header'`.

**Comment by perwezalamnewscorp** [comment]
I agree with you on this. If I add browser info using **reportInfo** then I expect to see report by browser name category

**For example**
if add  reportInfo "Chrome" , reportInfo "Firefox" . etc , then I should see all test report by browser category as

**Chrome**
Test1
Test2
..

**Firefox**
Test1
Test2
..
How could we achieve by using reportInfo ? . Because reportInfo can be used to add any kind of information into the report, but not necessarily it will be used to filter the test. But the test execution report must be segregated based on the browser name (I want this exactly).

**Comment by renatoathaydes** [comment]
Categorizing reports is not what this ticket is about, please create a separate ticket if you want to request that feature (though I would say it's out of scope, you should categorize your tests yourself by separating them in the appropriate manner in my opinion).

**Comment by perwezalamnewscorp** [comment]
Okay I will create the new ticket for this and have discussion over there. 

I have another requirement related to browser console message (Info, Warning, Error) and want these to be added into the report.

I believe this info has to be captured programmatically and add the same using reportInfo. It can easily be captured by using setup or cleanup fixture, But how it could be highlighted in the report based on the error type?

**Comment by schhil** [comment]
@renatoathaydes: I tested the feature in the "next" branch. It is already a very useful solution, which works and is good enough for me, since I can report all information I want :-). So I would already be happy if you'd release this. When testing, I found, if the info string contains a carriage return character "\n", this will result to a printed out "<br>" in the test report (Well, why would one do this, it make no sence inn this case, but I just let you know).
I attached an example picture from my test result:
![picture1](https://user-images.githubusercontent.com/11996721/33883286-8ab3458c-df3b-11e7-9c17-027911371cda.png)

If you like to further improove the feature in the future, I would recommend to allow to add  key-value-pairs to the report as an alternative to simple strings, such as 

    reportInfo  ["Request" : "GET", "URL" : "http://myAdress/..."]
    reportInfo  ["Headers" : "[application/json;charset=UTF-8]\nAccept-Encoding=[gzip,deflate]\nClientName=[abc.gsa]"

which would then result to a report that could look like this:
![picture2](https://user-images.githubusercontent.com/11996721/33883329-b76c68ce-df3b-11e7-8b76-d0854aba69ca.png)
So that report would basically contain the same information, but look even nicer, due to a tabular format.

**Comment by perwezalamnewscorp** [comment]
Wow. Its looking great and I like the idea of key-pair info is great. This info is all about the individual test. How can we add the Test Report Title in the index page of the report then ?. Title must be there in the index and individual spec page only (We can say this is the report name). Otherwise all info looks really cool in the snapshot

**Comment by renatoathaydes** [comment]
@schhil I think this feature has to be as general as possible, so I allow any text to be provided, including HTML text... so that people can just create their own "extensions" which display extra information in the reports in the manner that they want it to be, which I think would be impossible for Spock-Reports to do well enough for all user cases.

In your case, for example, I would create a Groovy method that creates the desired HTML, something like this:

```groovy
def tabularReportData(Map map) {
   reportInfo map.collect { k, v ->
        "<span>$k</span><span>$v</span>"
    }.join("<hr/>")
}
```

And then use it from a report like this:

```groovy
tabularReportData( ["Request" : "GET", "URL" : "http://myAdress/..."] )
```

**Comment by renatoathaydes** [comment]
@schhil added `reportHeader` method that lets you add information on the top of the report.... Also fixed the multi-report-creators issue (so this feature works even if you create several reports) and stopped escaping the text added to the report (I didn't know it was being escaped, but now it's not anymore), so you can do this in your reports:

```groovy
def setupSpec() {
    reportHeader "<strong>Browser under test: ${driver.browser.name}</strong>"
}
```

**Comment by renatoathaydes** [comment]
Feature released in 1.4.0. Thanks guys for all the input!
